### PR TITLE
Adding conv transpose 1d & 3d

### PR DESCRIPTION
**Pull request by Shashi456**
I'll add the tests. but till then could someone check the logic once and tell me if i'm wrong or right?

### CODE REVIEW

**Comment by Shashi456** [comment]
I've added 1d transpose as well, this is something not found in the keras library. But could someone check the logic of that as well? I'll add the rest of the conv layers throughout the week. 
I'll start writing test cases now

**Comment by Shashi456** [comment]
Note: The build passes.

**Comment by Shashi456** [comment]
@rxwei Any updates on taking a look at this?

**Comment by Shashi456** [comment]
@rxwei There seems to some disconnect between the tensorflow/Keras Convtranspose APIs and our APIs, could you help me write some tests for this?

File: Tests/TensorFlowTests/LayerTests.swift [hunk 1]
```diff
@@ -52,6 +52,40 @@ final class LayerTests: XCTestCase {
         XCTAssertEqual(output, expected)
     }
 
+    func testTransposedConv1D() {
+        let filter =  Tensor(shape: [1, 4, 1, 1], scalars: (0..<4).map(Float.init))
+        let bias = Tensor<Float>([8])
+        let layer = Conv1D<Float>(filter: filter, bias: bias, activation: identity, stride: 1, padding: .valid)
```
Comment (rxwei) [inline Tests/TensorFlowTests/LayerTests.swift]: Please fit all lines within 100 columns.

**Comment by rxwei** [comment]
> @rxwei There seems to some disconnect between the tensorflow/Keras Convtranspose APIs and our APIs, could you help me write some tests for this?

We will try to take a look later this week!

**Comment by Shashi456** [comment]
@rxwei Updated this PR, the build and tests pass on the latest nightly.

File: Sources/TensorFlow/Layers/Convolutional.swift [hunk 2]
```diff
@@ -450,6 +545,106 @@ public extension TransposedConv2D {
...
+
+    /// Returns the output obtained from applying the layer to the given input.
+    ///
+    /// - Parameter input: The input to the layer.
+    /// - Returns: The output.
+    @differentiable
+    public func callAsFunction(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
+        let batchSize = input.shape[0]
+        let w = (input.shape[1] - (1 * paddingIndex)) *
+          strides.0 + (filter.shape[0] * paddingIndex)
```
Comment (rxwei) [inline Sources/TensorFlow/Layers/Convolutional.swift]: Use 4-space indentation for line breaks.

File: Tests/TensorFlowTests/LayerTests.swift [hunk 3]
```diff
@@ -197,6 +197,29 @@ final class LayerTests: XCTestCase {
...
+        let expected = Tensor<Float>(shape: [1, 1, 4, 1],
+                                     scalars: [8, 9, 12, 18])
+        XCTAssertEqual(output, expected)
+    }
+
+    func testTransposedConv3D() {
+        let filter =  Tensor(shape: [2, 2, 2, 1, 1], scalars: (0..<8).map(Float.init))
+        let bias = Tensor<Float>([8])
+        let layer = TransposedConv3D(filter: filter, bias: bias, activation: identity,
+                                  strides: (1, 1, 1), padding: .same)
```
Comment (rxwei) [inline Tests/TensorFlowTests/LayerTests.swift]: Fix indentation.

File: Tests/TensorFlowTests/LayerTests.swift [hunk 4]
```diff
@@ -197,6 +197,29 @@ final class LayerTests: XCTestCase {
         XCTAssertEqual(output, expected)
     }
 
+    func testTransposedConv1D() {
+        let filter =  Tensor(shape: [4, 1, 1], scalars: (0..<4).map(Float.init))
+        let bias = Tensor<Float>([8])
+        let layer = TransposedConv1D(filter: filter, bias: bias, activation: identity, stride: 1, padding: .same)
```
Comment (rxwei) [inline Tests/TensorFlowTests/LayerTests.swift]: Fit within 100 columns.

File: Tests/TensorFlowTests/LayerTests.swift [hunk 5]
```diff
@@ -746,6 +769,8 @@ final class LayerTests: XCTestCase {
         ("testConv2DGradient", testConv2DGradient),
         ("testConv2DDilation", testConv2DDilation),
         ("testConv3D", testConv3D),
+        ("testTransposedConv1D",testTransposedConv1D),
```
Comment (rxwei) [inline Tests/TensorFlowTests/LayerTests.swift]: Add a space after each comma.

File: Sources/TensorFlow/Layers/Convolutional.swift [hunk 6]
```diff
@@ -351,6 +351,101 @@ public extension Conv3D {
...
+
+    /// Returns the output obtained from applying the layer to the given input.
+    ///
+    /// - Parameter input: The input to the layer.
+    /// - Returns: The output.
+    @differentiable
+    public func callAsFunction(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
+        let batchSize = input.shape[0]
+        let w = (input.shape[1] - (1 * paddingIndex)) *
+          stride + (filter.shape[0] * paddingIndex)
```
Comment (rxwei) [inline Sources/TensorFlow/Layers/Convolutional.swift]: 4-space indentation.

**Comment by Shashi456** [comment]
done

**Comment by Shashi456** [comment]
This can be merged

**Comment by Shashi456** [comment]
@rxwei could you approve this

File: Sources/TensorFlow/Layers/Convolutional.swift [hunk 7]
```diff
@@ -450,6 +545,106 @@ public extension TransposedConv2D {
...
+    public var filter: Tensor<Scalar>
+    /// The bias vector.
+    public var bias: Tensor<Scalar>
+    /// The element-wise activation function.
+    @noDerivative public let activation: Activation
+    /// The strides of the sliding window for spatial dimensions.
+    @noDerivative public let strides: (Int, Int, Int)
+    /// The padding algorithm for convolution.
+    @noDerivative public let padding: Padding
+    @noDerivative public let paddingIndex: Int
```
Comment (rxwei) [inline Sources/TensorFlow/Layers/Convolutional.swift]: Add a doc comment for this property.

File: Sources/TensorFlow/Layers/Convolutional.swift [hunk 8]
```diff
@@ -351,6 +351,101 @@ public extension Conv3D {
...
+    public var filter: Tensor<Scalar>
+    /// The bias vector.
+    public var bias: Tensor<Scalar>
+    /// The element-wise activation function.
+    @noDerivative public let activation: Activation
+    /// The strides of the sliding window for spatial dimensions.
+    @noDerivative public let stride: Int
+    /// The padding algorithm for convolution.
+    @noDerivative public let padding: Padding
+    @noDerivative public let paddingIndex: Int
```
Comment (rxwei) [inline Sources/TensorFlow/Layers/Convolutional.swift]: Add a doc comment for this property.

File: Tests/TensorFlowTests/LayerTests.swift [hunk 9]
```diff
@@ -113,28 +113,28 @@ final class LayerTests: XCTestCase {
...
+        let layer = Conv2D<Float>(filter: filter,
+                                  bias: bias,
                                   activation: identity,
-                                  strides: (2, 2), 
+                                  strides: (2, 2),
                                   padding: .valid)
         let input = Tensor(shape: [2, 4, 4, 2], scalars: (0..<64).map(Float.init))
         let grads = gradient( at: input, layer) { $1($0).sum() }
-        // The expected gradients were computed using the following Python code: 
+        // The expected gradients were computed using the following Python code:
```
Comment (rxwei) [inline Tests/TensorFlowTests/LayerTests.swift]: #499

Comment (rxwei) [inline Tests/TensorFlowTests/LayerTests.swift]: #499

Comment (Shashi456) [reply Tests/TensorFlowTests/LayerTests.swift]: This is already taken care of by #502, I'll pull into master.

File: Sources/TensorFlow/Layers/Convolutional.swift (hunk 7 above)
Comment (Shashi456) [reply Sources/TensorFlow/Layers/Convolutional.swift]: @rxwei this property isn't really exposed to the users, it's a temporary property which allows us to handle computation based on what kind of padding it is. we currently support `.same` and `.valid`.
Do we still document this?

**Comment by burmako** [comment]
@Shashi456 I've been assigned to help move this PR over the finish line. At a glance, things are looking great :)

If you could resolve the conflicts and add documentation comments for `paddingIndex` (something that explains why those fields are necessary), I'll run the tests and then we'll merge. Thank you!

**Comment by Shashi456** [comment]
@burmako this property isn't really exposed to the users, it's a temporary property which allows us to handle computation based on what kind of padding it is. we currently support .same and .valid.
Do we still document this? 
When we support the third  padding feature, i think this will be omitted.

**Comment by burmako** [comment]
@Shashi456 I think a comment along the following lines would be great:

```
/// Temporary property which allows us to handle computation based on padding.
/// Will be removed when issue #1234 is fixed.
```

**Comment by Shashi456** [comment]
@burmako does this work?

**Comment by burmako** [comment]
Thank you! I'll be merging now :)
